### ISSUE TITLE
Consider using ex-info for exceptions

### ISSUE THREAD

**Issue by anderseknert** [johanfylling/jarl#65]
Using [ex-info](https://clojuredocs.org/clojure.core/ex-info) seems to be the idiomatic way of working with exceptions in code that targets both Clojure and Clojurescript. The "type" of the exception is provided in the form of metadata rather than classes, i.e:

```clojure
(defn something [x]
  (if (= x 1)
    x
    (throw (ex-info "exception message here"
                    {:type :builtin-exception
                     :other-metadata "here"}
                    (comment "cause here"))))
```

I think we should consider this approach for internal code, while converting them to "proper" Java exceptions only when throwing something expected to be caught by users of the Java API.

**Comment by johanfylling** [comment]
We'll need to deal with exceptions thrown by custom built-ins implemented in java. Might be as simple as just transforming them to this structure; but it'd be nice if we could keep the `cause` chain and stack trace, for when the exception is propagated all the way out to the library caller.

Perhaps we can piggy-back the original exception on a field, or keep the evaluator agnostic to the actual types, and use helper-functions for throwing and catching exceptions.

**Comment by anderseknert** [comment]
For sure! That seems like an integration point where it makes sense to use/check for whatever the API client expects. Those places should however be quite few, and hopefully easy to isolate. This is mainly just for how we raise and deal with exceptions internally 🙂

**Comment by anderseknert** [comment]
Fixed in #79
